src/wavm.cpp

```diff
@@ -0,0 +1,194 @@
+  bool meterInterfaceGas
+) {
+  HERA_DEBUG << "Executing with wavm...\n";
+
+  // set up a new ethereum interface just for this contract invokation
+  ExecutionResult result;
+  WavmEthereumInterface myInterface{context, state_code, msg, result, meterInterfaceGas};
+  wavm_host_module::interface.push(&myInterface);
+
+  // first parse module
+  std::vector<U8> codeBytes = static_cast< std::vector<U8> > (code);
+  IR::Module moduleAST; 
+  bool loadedSuccess = loadBinaryModule(codeBytes.data(), codeBytes.size(), moduleAST);
+  heraAssert(loadedSuccess, "wavm couldn't parse module into syntax tree");
+
+  // next set up the host module. Note: in ewasm, we create a new VM for each call to a module, so we must instantiate a new host module for each of these VMs, this is inefficient, but OK for prototyping.
+  // compartment is like the Wasm store, represents the VM, has lists of globals, memories, tables, and also has wavm's runtime stuff
+  Runtime::GCPointer<Runtime::Compartment> compartment = Runtime::createCompartment();
+  // context stores the compartment and some other stuff
+  Runtime::GCPointer<Runtime::Context> wavm_context = Runtime::createContext(compartment);
+  // instantiate host Module
+  HashMap<std::string, Runtime::Object*> extraEthereumExports; //empty for current ewasm stuff
+  Runtime::GCPointer<Runtime::ModuleInstance> ethereumHostModule = Intrinsics::instantiateModule(compartment, wavm_host_module::INTRINSIC_MODULE_REF(ethereum), "ethereum", extraEthereumExports);
+  heraAssert(ethereumHostModule, "wavm couldn't instantiate host module");
+  // prepare contract module to resolve links against host module
+  wavm_host_module::HeraWavmResolver myResolver(compartment);
+  myResolver.moduleNameToInstanceMap.set("ethereum", ethereumHostModule);
+  Runtime::LinkResult linkResult = linkModule(moduleAST, myResolver);
+  heraAssert(linkResult.success, "wavm couldn't link contract against host module");
+
+  // instantiate contract module
+  Runtime::GCPointer<Runtime::ModuleInstance> moduleInstance = instantiateModule( compartment, moduleAST, std::move(linkResult.resolvedImports), "name of ewasm contract");
+  heraAssert(moduleInstance, "wavm couldn't instantiate contract module");
+
+  // get memory for easy access in host functions
+  wavm_host_module::interface.top()->setWasmMemory( asMemory(getInstanceExport(moduleInstance, "memory")) );
+
+  // invoke the main function
+  Runtime::GCPointer<Runtime::FunctionInstance> functionInstance = asFunctionNullable(getInstanceExport(moduleInstance, "main"));
+  heraAssert(functionInstance, "wavm couldn't find main function");
+  std::vector<IR::Value> invokeArgs; //empty for current ewasm stuff
+  Runtime::catchRuntimeExceptions(  //this is how WAVM's try/catch for exceptions
+    [&] {
+      try { //this is how Hera's try/catch for exceptions
+        Runtime::invokeFunctionChecked(wavm_context, functionInstance, invokeArgs);
+      } catch (EndExecution const&) {
```

Comment:
What about the other exceptions? Are they passed through?

Reply:
Both WAVM and Hera exceptions seem to be handled properly for all test cases for opcodes `useGas`, `getCallDataSize`, `callDataCopy`, `revert`, `finish`.

Try it yourself, all of the test cases are `useGas`, `getCallDataSize`, `callDataCopy`, `callDataCopy256`, `revert`, `revertVeryLong`, `returnPredefinedData`, and `returnVeryLong`.